File: 数据结构与算法/2.数组/lua/array.lua
local ArrayList = {
  mCap = 0,
  mData = {},
  mLength = 0,
}

function ArrayList:New(capacity)
  o = {}
  self.mCap = capacity
  self.mData = {}
  self.mLength = 0

  setmetatable(o,{ __index = self })
  return o
end

function ArrayList:Insert(index, data)
  assert(type(index) == "number", "index is not a int")
  -- assert(self.mCap == self.mData, "数组没有多余空间进行添加") 还是去扩容算了
  assert(index >= 1 and index <= self.mLength+1, "Index was outside the bounds of the list")
  if(self.mCap == self.mLength)
  then
    self:resize()
  end
  -- lua 的for循环感觉不太适合条件遍历呀......有点难把控
  -- for i = self.mLength+1, self.mLength-index+2, -1 do
  --   print("self.mData["..i.."]:"..tostring(self.mData[i]))
  --   print("self.mData["..(i+1).."]:"..tostring(self.mData[i+1]))
  --   self.mData[i] = self.mData[i-1]
  -- end
  --lua中的下标从1开始.......
  local i = self.mLength+1
  while(i > index)
  do  
    self.mData[i] = self.mData[i-1]
    i = i - 1
  end

  self.mData[index] = data
  self.mLength = self.mLength + 1

  return true
end

function ArrayList:Find(index)
  assert(type(index) == "number", "index is not a number")
  assert(index >= 1 and index < self.mLength+1, "Index was outside the bounds of the list")

  -- result["result"] = true
  -- result["data"] = self.mData[index]
  print(self.mData[index])

  return true
end

function ArrayList:Delete(index)
  assert(type(index) == "number", "index is not a int")
  assert(index >= 1 and index <= self.mLength+1, "Index was outside the bounds of the list")

  data = self.mData[index]
  while(self.mData[index] ~= nil)
  do
    self.mData[index] = self.mData[index+1]
    index = index+1
  end

  return true
end

function ArrayList:resize()
  self.mCap = self.mCap * 2
end


-- 测试
local a = ArrayList:New(8)
-- local b =  ArrayList:New(16)
--初始容量
print(a.mCap)
--插入10个数字，且容量进行扩充
a:Insert(1,2)
a:Insert(2,3)
a:Insert(3,4)
a:Insert(4,5)
a:Insert(5,6)
a:Insert(6,7)
a:Insert(7,8)
a:Insert(8,9)
a:Insert(9,10)
a:Insert(10,11)

--查找第三个数 输出4
a:Find(3)

--删除第三个数
a:Delete(3)

--原本第四个数替换成第三个数 输出5
a:Find(3)

--插入第三位数
a:Insert(3,50)

--再次查找第三位数，显示50
a:Find(3)
--查找第四位，之前的第三位再次变为第四位
a:Find(4)
--查找第10位，输出11
a:Find(10)

--进行了扩容，当前容量为16
print(a.mCap)

print("index of a list:")
for i,v in ipairs(a.mData) do
  print(v)  
end

-- print("index of b list:")
-- for i,v in ipairs(b.mData) do
--   print(v)  
-- end

File: DataStructGame/Assets/Game/Scripts/CSharp/Array/Array.cs
using System;
using Game.Scripts.Common.CSharp;
using UnityEngine;

namespace Game.Scripts.CSharp.Array
{
   /**描述：c#自定义数组
  功能点：
       1. 可以动态扩容
       2. 可以插入，删除
       3. 本身自带快速随机访问特性，因此不特地提供查询
  */
  public class Array<T> {
      private const int DEFAULT_CAPACITY = 10;
      
      private T[] m_data;
      private int m_capacity;
      private int m_length               = 0;
      public T[] data   => m_data;
      public int capacity => m_capacity;
      public int length => m_length;

      public T this[int index] {
        get {
          //index 不可以与 m_length 相等，否则超出下标
          if (Common.CSharp.Common.CheckIndexOutForEqual(index, m_length))
            throw new Exception("数组出现越界");
          return m_data[index];
        }
        set {
          //设定只能重新赋值，不能进行添加，否则需要加入扩容操作
          if (Common.CSharp.Common.CheckIndexOutForEqual(index, m_length))
            return;
          m_data[index] = value;
        }
      }
      
      public Array() : this(DEFAULT_CAPACITY){
        
      }

      public Array(int size) {
        _Resize(size);
      }

      public bool Insert(int index, T new_Elem){
        //判断下标是否正确，防止出现不连续下标以及负数下标
        if (Common.CSharp.Common.CheckIndexOutForNotEqual(index, m_length))
          return false;
        
        //判断是否有容量
        if(m_length >= m_capacity){
          _Resize(m_capacity*2);
        }

        //此处把替换位置后面的值进行平移挪动，然后进行赋值
        for(int i = m_length; i > index; i--){
          m_data[i] = m_data[i-1];
        }

        m_data[index] = new_Elem;
        m_length++;

        return true;
      }
      
      public bool Delete(int index){
        //判断下标是否正确，防止出现不连续下标以及负数下标
        if(Common.CSharp.Common.CheckIndexOutForEqual(index,m_length))
          return false;
        
        //此处把替换位置前面的值进行平移挪动，然后进行赋值
        for(int i = index; i < m_length-1; i++){
          m_data[i] = m_data[i+1];
        }

        m_length--;

        return true;
      }

      public void Clear(){
        m_data     = null;
        m_length   = 0;
        m_capacity = 0;
      }

      public void PrintAll(){
        for(int i = 0; i < m_length; i++){
          Debug.Log("mData["+i+"]:"+m_data[i].ToString());
        }
      }
      
      //双倍扩容
      private void _Resize(int size){
        m_capacity = size;
        T[] newData = new T[m_capacity];

        for(int i = 0; i < m_length; i++){
          newData[i] = m_data[i];
        }

        m_data = newData;
      }

   } 
}